adrs/0001-cli-design.md

```diff
@@ -0,0 +1,306 @@
+* Given a CVE ID, what components are affected?
+* What products + version + stream contain a given component (e.g. full text search)?
+* What are the fixed CVEs for a product + version + stream?
+* What are the fixed CVEs for a component?
+* What are the won’t fix CVEs for a component?
+* What are the won’t fix CVEs for a product?
+* How many CVE’s are filed against a product + version
+
+For example, to get a list of cves affecting a specific component:
+```commandline
+> griffon queries component_cves --purl "pkg:rpm/redhat/curl@7.29.0-19.el7?arch=aarch64" --affectedness AFFECTED 
+ ```
+
+A set of 'canned' queries masks the complexity of calling up individual services and synthesizing results manually.
+
+#### Process operations
+
+A **process** represents business logic typically resulting in an update of state on service entities.
+
+### Manage operations
+A separate entrypoint for management operations is provided as convenience.
+
+In the short term, this area is meant for internal usage for managing core data services.
+
+```commandline
+> griffon manage osidb status 
+> griffon manage corgi status
+```
+
+### Plugins
+
+We **MAY** develop an extensible set of plugins to external services as an 'incubation' area for 
+integrating with third party services. This gives the CLI a good 'swiss army knife' utility as
+well as help identify integrations with more query/processes leveraging 3rd party services.
+
+A few example plugins are provided, for example, an example plugin integrating with https://osv.dev:
+
+```commandline
+> griffon z_osv query-by-commit-hash --commit_hash 6879efc2c1596d11a6a6ad296f80063b558d5e0f
+
+```
+Plugins are dynamically 'registered' once they are copied into the griffon/commands/plugins folder
+allowing individual end users to develop their own customisations.
+
+## Autocompletion
+
+The python click module [provides](https://click.palletsprojects.com/en/8.1.x/shell-completion/) provides custom shell completion.
+
+We **MAY** enable live autocompletion of data such as CVE-ids, product stream names/ofuris making it easier to use
+griffon.
+
+This can be achieved by generation of python module containing static values (ex. griffon/autocomplete/product_stream.py)
+or enhancing data services to provide autocompletion endpoints.
+
+## Process flow
+
+A lightweight process flow is defined as follows:
+
+1. CLI validate options/input
+2. invoke operation
+3. transform and emit output
+
+## Enhance services
+
+Some considerations for enhancing service bindings:
+
+* 'bake in' concurrent service requests (using threads or asyncio)
+* component-registry bindings need to be deployed on github and published to PyPI
+* component-registry bindings components.retrieve_list needs to support **ofuri**
+* retrieve a batch set of components with a list of {uuid|nvr|purl}
+* exception handling in bindings
+* implement next() in both bindings
+
+In addition, OSIDB needs to implement a 2nd affect type for mapping a single purl 
```

Comment:
This is definitely a valid and good point. It would be of course nice to have this available soon. The hard question is however how would it fit together with the current Bugzilla data format + SFM2 would definitely not understand it. Therefore it is IMO a post-SFM2 thing and possibly even post-Bugzilla thing.

Reply:
in the short term there is nothing stopping griffon crafting old world affects (like depcli does) for local sfm2 ...we will walk this discussion into OSIDB team.